### PR TITLE
Interactivity API: Server Directive Processor for `data-wp-each`

### PR DESCRIPTION
**Pull request by luisherranz**
Tracking Issue: https://github.com/WordPress/gutenberg/issues/56803

## What?

This PR adds an additional server directive processor to process the `data-wp-each` directive, introduced in:

- https://github.com/WordPress/gutenberg/pull/57859

## Why?

Because all the directives that modify the markup should be processed in the server to avoid layout shifts.

## How?

By introducing a new method to append content after the closing tag of balanced tags (or after a void tag) and refactoring the `process_directive` method to be used internally with the inner content of the `template` that contains the `data-wp-each` directive. 

It also makes the `next_balanced_tag_closer_tag` public (but with `@access private`) to be able to jump from top-level nodes and add the corresponding `data-wp-each-child` directives, which are required during the frontend hydration.

This PR also includes a `compat` version of the fallback of the native PHP8 `array_is_list` function which is going to be introduced in WP 6.5.

## Testing Instructions

- Create a block that has a `data-wp-each` directive and pass a reference to an array.
- Render that block in the frontend and inspect the HTML received from the server to make sure that the tags contained inside the `template` tag are properly rendered.

### CODE REVIEW

File: phpunit/interactivity-api/class-wp-interactivity-api-directives-processor-test.php [hunk 1]
```diff
@@ -366,4 +366,354 @@ public function test_is_void_element() {
...
+	 *
+	 * @covers ::next_balanced_tag_closer_tag
+	 */
+	public function test_next_balanced_tag_closer_tag_standard_tags() {
+			$content = '<div>Text</div>';
+			$p       = new WP_Interactivity_API_Directives_Processor( $content );
+			$p->next_tag();
+			$this->assertTrue( $p->next_balanced_tag_closer_tag() );
+			$this->assertEquals( 'DIV', $p->get_tag() );
+			$this->assertTrue( $p->is_tag_closer() );
```
Comment (DAreRodz) [inline phpunit/interactivity-api/class-wp-interactivity-api-directives-processor-test.php]:
```suggestion
		$content = '<div>Text</div>';
		$p       = new WP_Interactivity_API_Directives_Processor( $content );
		$p->next_tag();
		$this->assertTrue( $p->next_balanced_tag_closer_tag() );
		$this->assertEquals( 'DIV', $p->get_tag() );
		$this->assertTrue( $p->is_tag_closer() );
```

It seems there's an extra tab in each line.

PS: Same in other tests below. 👇

File: phpunit/interactivity-api/class-wp-interactivity-api-directives-processor-test.php [hunk 2]
```diff
@@ -366,4 +366,354 @@ public function test_is_void_element() {
...
+	public function test_next_balanced_tag_closer_tag_no_matching_closing_tag() {
+			$content = '<div>No closing tag here';
+			$p       = new WP_Interactivity_API_Directives_Processor( $content );
+			$p->next_tag();
+
+			$content = '<div><div>No closing tag here</div>';
+			$p       = new WP_Interactivity_API_Directives_Processor( $content );
+			$p->next_tag();
+			$this->assertFalse( $p->next_balanced_tag_closer_tag() );
+		$this->assertFalse( $p->next_balanced_tag_closer_tag() );
```
Comment (DAreRodz) [inline phpunit/interactivity-api/class-wp-interactivity-api-directives-processor-test.php:686]: This line probably was intended to be after 696. ☝️

File: phpunit/interactivity-api/class-wp-interactivity-api-wp-each-test.php [hunk 3]
```diff
@@ -0,0 +1,516 @@
...
+		$original = '' .
+			'<template data-wp-each="myPlugin::state.list">' .
+				'<span data-wp-text="myPlugin::context.item"></span>!' .
+			'</template>';
+		$new      = $this->interactivity->process_directives( $original );
+		$this->assertEquals( $original, $new );
+	}
+
+	/**
+	 * Tests that the `data-wp-each` directive works with nestded template tags.
```
Comment (DAreRodz) [inline phpunit/interactivity-api/class-wp-interactivity-api-wp-each-test.php]:
```suggestion
	 * Tests that the `data-wp-each` directive works with nested template tags.
```

**Review by DAreRodz — APPROVED** [review]
The code and tests look great. 👏

PS: I've left some nitpicky suggestions.

File: phpunit/interactivity-api/class-wp-interactivity-api-directives-processor-test.php (hunk 1 above)
Comment (luisherranz) [reply phpunit/interactivity-api/class-wp-interactivity-api-directives-processor-test.php]: Thanks! Fixed in https://github.com/WordPress/gutenberg/pull/58498/commits/72a2f19de85622bfa4274aa5074776855c3b38c7.

File: phpunit/interactivity-api/class-wp-interactivity-api-wp-each-test.php (hunk 3 above)
Comment (luisherranz) [reply phpunit/interactivity-api/class-wp-interactivity-api-wp-each-test.php]: Thanks! Fixed in https://github.com/WordPress/gutenberg/pull/58498/commits/72a2f19de85622bfa4274aa5074776855c3b38c7.

File: phpunit/interactivity-api/class-wp-interactivity-api-directives-processor-test.php (hunk 2 above)
Comment (luisherranz) [reply phpunit/interactivity-api/class-wp-interactivity-api-directives-processor-test.php:686]: Thanks! Fixed in https://github.com/WordPress/gutenberg/pull/58498/commits/72a2f19de85622bfa4274aa5074776855c3b38c7.

**Comment by luisherranz** [comment]
Thanks David! Feedback addressed.

**Comment by luisherranz** [comment]
Let's port this to core and continue working there:

- https://github.com/WordPress/wordpress-develop/pull/5953/

**Pull request merged**